File: src/WagsMediaRepository.Domain/Models/Book.cs
namespace WagsMediaRepository.Domain.Models;

public class Book
{
    public int BookId { get; set; }
    
    public int BookStatusId { get; set; }
    
    public int BookTypeId { get; set; }
    
    public int? BookSeriesId { get; set; }

    public string Title { get; set; } = string.Empty;
    
    public string SubTitle { get; set; } = string.Empty;

    public string FullTitle
    {
        get
        {
            if (string.IsNullOrWhiteSpace(SubTitle))
            {
                return Title;
            }

            return $"{Title}: {SubTitle}";
        }
    }
    
    public string Author { get; set; } = string.Empty;
    
    public string Link { get; set; } = string.Empty;
    
    public DateTime? DateStarted { get; set; }
    
    public DateTime? DateCompleted { get; set; }
    
    public int Rating { get; set; }

    public string BookNotesUrl { get; set; } = string.Empty;
    
    public string Thoughts { get; set; } = string.Empty;
    
    public string CoverImageUrl { get; set; } = string.Empty;
    
    public bool IsAtLibrary { get; set; }
    
    public bool IsPurchased { get; set; }

    public int CurrentPage { get; set; } = 1;

    public int PageCount { get; set; }

    public decimal PercentComplete
    {
        get
        {
            if (PageCount == 0)
            {
                return 0;
            }

            return Math.Round(((decimal)CurrentPage / (decimal)PageCount) * 100);
        }
    }
    
    public int? SortOrder { get; set; }

    public BookStatus Status { get; set; } = new();
    
    public BookType Type { get; set; } = new();
    
    public BookSeries? Series { get; set; }

    public List<BookGenre> Genres { get; set; } = [];
    
    public List<BookFormat> Formats { get; set; } = [];

    public static Book FromDto(BookDto dto) => new()
    {
        BookId = dto.BookId,
        BookStatusId = dto.BookStatusId,
        BookTypeId = dto.BookTypeId,
        BookSeriesId = dto.BookSeriesId,
        Title = dto.Title,
        SubTitle = dto.SubTitle,
        Author = dto.Author,
        Link = dto.Link,
        DateStarted = dto.DateStarted,
        DateCompleted = dto.DateCompleted,
        Rating = dto.Rating,
        BookNotesUrl = dto.BookNotesUrl,
        Thoughts = dto.Thoughts,
        CoverImageUrl = dto.CoverImageUrl,
        IsAtLibrary = dto.IsAtLibrary,
        IsPurchased = dto.IsPurchased,
        CurrentPage = dto.CurrentPage,
        PageCount = dto.PageCount,
        SortOrder = dto.SortOrder,
        Status = BookStatus.FromDto(dto.BookStatus),
        Type = BookType.FromDto(dto.BookType),
        Series = dto.BookSeries is not null ? BookSeries.FromDto(dto.BookSeries) : null,
        Genres = dto.BookToBookGenres.Select(bg => BookGenre.FromDto(bg.BookGenre)).ToList(),
        Formats = dto.BookToBookFormats.Select(bf => BookFormat.FromDto(bf.BookFormat)).ToList(),
    };
}

File: api/src/models/book.ts
export type BookSeries = {
	bookSeriesId: number;
	name: string;
	colorCode: string;
	bookCount?: number;
}

export type BookGenre = {
	bookGenreId: number;
	name: string;
	colorCode: string;
	bookCount?: number;
}

export type BookFormat = {
	bookFormatId: number;
	name: string;
	colorCode: string;
	bookCount?: number;
}

export type BookStatus = {
	bookStatusId: number;
	name: string;
	colorCode: string;
}

export type BookType = {
	bookTypeId: number;
	name: string;
	colorCode: string;
}

export type Book = {
	bookId: number;
	bookStatusId: number;
	bookTypeId: number;
	bookSeriesId?: number;
	title: string;
	subTitle: string;
	fullTitle?: string;
	author: string;
	link: string;
	dateStarted?: Date | string | null;
	dateCompleted?: Date | string | null;
	rating: number;
	thoughts: string;
	bookNotesUrl?: string;
	coverImageUrl: string;
	currentPage: number;
	pageCount: number;
	progress: number;
	sortOrder?: number;
	isAtLibrary: boolean;
	isPurchased: boolean;
	heardAboutFrom: string;
	genres: BookGenre[];
	formats: BookFormat[];
	status?: BookStatus;
	type?: BookType;
	series?: BookSeries;
}

export type BookGenreLink = {
	bookId: number;
	genreId: number;
	genreName: string;
	genreColorCode: string;
}

export type BookFormatLink = {
	bookId: number;
	formatId: number;
	fomatName: string;
	formatColorCode: string;
}

export type BookQueryReturn = {
	BookId: number;
	BookStatusId: number;
	BookTypeId: number;
	BookSeriesId: number;
	Title: string;
	SubTitle: string;
	Author: string;
	Link: string;
	DateStarted?: Date;
	DateCompleted?: Date;
	Rating: number;
	Thoughts: string;
	BookNotesUrl: string;
	CoverImageUrl: string;
	CurrentPage: number;
	PageCount: number;
	Progress: number;
	SortOrder: number;
	IsAtLibrary: boolean;
	IsPurchased: boolean;
	BookStatusName: string;
	BookStatusColor: string;
	BookTypeName: string;
	BookTypeColor: string;
	BookSeriesName?: string;
	BookSeriesColor?: string;
	HeardAboutFrom: string;
}

export type BookGenreLinkQueryReturn = {
	BookId: number;
	BookGenreId: number;
	BookGenreName: string;
	BookGenreColor: string;
}

export type BookFormatLinkQueryReturn = {
	BookId: number;
	BookFormatId: number;
	BookFormatName: string;
	BookFormatColor: string;
}

export type BookGenreQueryReturn = {
	BookGenreId: number;
	Name: string;
	ColorCode: string;
	BookCount: number;
}

export type BookFormatQueryReturn = {
	BookFormatId: number;
	Name: string;
	ColorCode: string;
	BookCount: number;
}

export type BookSeriesQueryReturn = {
	BookSeriesId: number;
	Name: string;
	ColorCode: string;
	BookCount: number;
}